scripts/config_test_vm.psm1

```diff
@@ -0,0 +1,238 @@
+﻿# Copyright (c) Microsoft Corporation
+# SPDX-License-Identifier: MIT
+
+param ([Parameter(Mandatory=$True)] [string] $Admin,
+       [Parameter(Mandatory=$True)] [SecureString] $AdminPassword,
+       [Parameter(Mandatory=$True)] [string] $WorkingDirectory,
+       [Parameter(Mandatory=$True)] [string] $LogFileName)
+
+
+Push-Location $WorkingDirectory
+
+Import-Module .\common.psm1 -Force -ArgumentList ($LogFileName) -WarningAction SilentlyContinue
+
+#
+# VM Initialization functions
+#
+
+function Wait-AllVMsToInitialize
+{
+    param([Parameter(Mandatory=$True)]$VMList,
+          [Parameter(Mandatory=$True)][string] $UserName,
+          [Parameter(Mandatory=$True)][SecureString] $AdminPassword)
+
+    $totalSleepTime = 0
+    $ReadyList = @{}
+    do {
+        foreach ($VM in $VMList) {
+            $VMName= $VM.Name
+            if ($ReadyList[$VMName] -ne $True) {
+                $HeartBeat = (Get-VMIntegrationService $VMName | ?{$_.name -eq "Heartbeat"}).PrimaryStatusDescription
+                if ($HeartBeat -eq "OK") {
+                    $ReadyList += @{$VMName = $True}
+                } else {
+                    break
+                }
+                Write-Log "Heartbeat OK on $VMName" -ForegroundColor Green
+            }
+        }
+        if ($ReadyList.Count -ne $VMList.Count) {
+            Write-Log ("{0} of {1} VMs are ready." -f $ReadyList.Count, $VMList.Count)
+            #sleep for 30 seconds
+            Start-Sleep -seconds 30
+            $totalSleepTime += 30
+        }
+    }
+    until (($ReadyList.Count -eq $VMList.Count) -or ($totalSleepTime -gt 5*60))
+
+    if ($ReadyList.Count -ne $VMList.Count) {
+        throw ("Did not get heartbeat from one or more VMs 5 minutes after starting")
+    }
+
+    $TestCred = New-Credentials -Username $UserName -AdminPassword $AdminPassword
+    $ReadyList.Clear()
+    $totalSleepTime = 0
+
+    do {
+        foreach($VM in $VMList) {
+            $VMName = $VM.Name
+            if ($ReadyList[$VMName] -ne $True) {
+                Write-Log "Poking $VMName to see if it is ready to accept commands"
+                $ret = Invoke-Command -VMName $VMName -Credential $TestCred -ScriptBlock {$True} -ErrorAction SilentlyContinue
```

Comment:
Does this fail if the VM is not ready? Does it hang at all?

Reply:
No - times out after 5 minutes. I can reduce that if you prefer.